source/slang-glslang/slang-glslang.cpp

```diff
@@ -86,6 +89,119 @@ static void dumpDiagnostics(
     dump(log.c_str(), log.length(), request->diagnosticFunc, request->diagnosticUserData, stderr);
 }
 
+// Apply the SPIRV-Tools optimizer to generated SPIR-V based on the desired optimization level
+// TODO: add flag for optimizing SPIR-V size as well
+static void glslang_optimizeSPIRV(std::vector<unsigned int>& spirv, unsigned optimizationLevel, unsigned debugInfoType)
+{
+    spv_target_env target_env = SPV_ENV_UNIVERSAL_1_2;
+
+    spvtools::Optimizer optimizer(target_env);
+    optimizer.SetMessageConsumer(
+        [](spv_message_level_t level, const char *source, const spv_position_t &position, const char *message) {
+        auto &out = std::cerr;
+        switch (level)
+        {
+        case SPV_MSG_FATAL:
+        case SPV_MSG_INTERNAL_ERROR:
+        case SPV_MSG_ERROR:
+            out << "error: ";
+            break;
+        case SPV_MSG_WARNING:
+            out << "warning: ";
+            break;
+        case SPV_MSG_INFO:
+        case SPV_MSG_DEBUG:
+            out << "info: ";
+            break;
+        default:
+            break;
+        }
+        if (source)
+        {
+            out << source << ":";
+        }
+        out << position.line << ":" << position.column << ":" << position.index << ":";
+        if (message)
+        {
+            out << " " << message;
+        }
+        out << std::endl;
+    });
+
+    // If debug info is being generated, propagate
+    // line information into all SPIR-V instructions. This avoids loss of
+    // information when instructions are deleted or moved. Later, remove
+    // redundant information to minimize final SPRIR-V size.
+    if (debugInfoType != SLANG_DEBUG_INFO_LEVEL_NONE)
+    {
+        optimizer.RegisterPass(spvtools::CreatePropagateLineInfoPass());
+    }
+
+    // TODO confirm which passes we want to invoke for each level
+    switch (optimizationLevel)
+    {
+    case SLANG_OPTIMIZATION_LEVEL_NONE:
+        // Don't register any passes if our optimization level is none
+        break;
+    case SLANG_OPTIMIZATION_LEVEL_DEFAULT:
+        // Use a minimal set of performance settings
+        optimizer.RegisterPass(spvtools::CreateInlineExhaustivePass());
+        optimizer.RegisterPass(spvtools::CreateAggressiveDCEPass());
+        optimizer.RegisterPass(spvtools::CreatePrivateToLocalPass());
+        optimizer.RegisterPass(spvtools::CreateScalarReplacementPass(100));
+        optimizer.RegisterPass(spvtools::CreateLocalAccessChainConvertPass());
+        optimizer.RegisterPass(spvtools::CreateAggressiveDCEPass());
+        break;
+    case SLANG_OPTIMIZATION_LEVEL_HIGH:
+    case SLANG_OPTIMIZATION_LEVEL_MAXIMAL:
+        // Use the same passes when specifying the "-O" flag in spirv-opt
+        optimizer.RegisterPass(spvtools::CreateWrapOpKillPass());
```

Comment:
A possible minor cleanup for us to consider in the future is organizing this logic so that it is clear that the passes at each level are (usually) a subset of those we run at the higher levels, with the relative ordering preserved. This can be done with logic like:

```
if(optimizationLevel > 0) {
   addBasicOpt();
   ...
   if(optimizationLevel >= 2) {
       addMoreExpensiveOpt();
    }
    ...
}
```

That can make it easier to build and maintain the optimization passes in a consistent way across the optimization levels, so that you don't end up with weird cases where decreasing the optimization level gives better results becaues a lower level runs the passes in a different order (or some passes got added to the lower level and forgotten at the higher).

Reply:
This was how I initally wanted to do things as well, except I've noticed that there are a couple of places where the order that the passes are registered in seems to be important. For example there are several dead code elimination passes registered in between others. spirv-opt seems to have specific "profiles" attached to optimization parameters (speed, size, HLSL legalization, WebGL). But, I should definitely move the pass registration to helper functions to make the code look a little cleaner.